fix test for docsrs

https://github.com/linera-io/linera-protocol/actions/runs/5173690463/jobs/9319195346

```
Error: Building documentation failed

Caused by:
    Building documentation failed

Caused by:
    failed to load local package /home/runner/work/_temp/crates/linera-core/package/linera-core-0.1.0

Caused by:
    command failed: exit status: 101

Stack backtrace:
   0: <core::result::Result<T,F> as core::ops::try_trait::FromResidual<core::result::Result<core::convert::Infallible,E>>>::from_residual
             at /rustc/84c898d65adf2f39a5a98507f1fe0ce10a2b8dbc/library/core/src/result.rs:2065:27
   1: docs_rs::utils::cargo_metadata::CargoMetadata::load_from_rustwide
             at ./src/utils/cargo_metadata.rs:18:19
   2: docs_rs::docbuilder::rustwide_builder::RustwideBuilder::build_local_package
             at ./src/docbuilder/rustwide_builder.rs:316:24
   3: cratesfyi::BuildSubcommand::handle_args
             at ./src/bin/cratesfyi.rs:418:21
   4: cratesfyi::CommandLine::handle_args
             at ./src/bin/cratesfyi.rs:161:18
   5: cratesfyi::main
             at ./src/bin/cratesfyi.rs:65:23
   6: core::ops::function::FnOnce::call_once
             at /rustc/84c898d65adf2f39a5a98507f1fe0ce10a2b8dbc/library/core/src/ops/function.rs:250:5
   7: std::sys_common::backtrace::__rust_begin_short_backtrace
             at /rustc/84c898d65adf2f39a5a98507f1fe0ce10a2b8dbc/library/std/src/sys_common/backtrace.rs:134:18
   8: std::rt::lang_start::{{closure}}
             at /rustc/84c898d65adf2f39a5a98507f1fe0ce10a2b8dbc/library/std/src/rt.rs:166:18
   9: core::ops::function::impls::<impl core::ops::function::FnOnce<A> for &F>::call_once
             at /rustc/84c898d65adf2f39a5a98507f1fe0ce10a2b8dbc/library/core/src/ops/function.rs:287:13
      std::panicking::try::do_call
             at /rustc/84c898d65adf2f39a5a98507f1fe0ce10a2b8dbc/library/std/src/panicking.rs:487:40
      std::panicking::try
             at /rustc/84c898d65adf2f39a5a98507f1fe0ce10a2b8dbc/library/std/src/panicking.rs:451:19
      std::panic::catch_unwind
             at /rustc/84c898d65adf2f39a5a98507f1fe0ce10a2b8dbc/library/std/src/panic.rs:140:14
      std::rt::lang_start_internal::{{closure}}
             at /rustc/84c898d65adf2f39a5a98507f1fe0ce10a2b8dbc/library/std/src/rt.rs:148:48
      std::panicking::try::do_call
             at /rustc/84c898d65adf2f39a5a98507f1fe0ce10a2b8dbc/library/std/src/panicking.rs:487:40
      std::panicking::try
             at /rustc/84c898d65adf2f39a5a98507f1fe0ce10a2b8dbc/library/std/src/panicking.rs:451:19
      std::panic::catch_unwind
             at /rustc/84c898d65adf2f39a5a98507f1fe0ce10a2b8dbc/library/std/src/panic.rs:140:14
      std::rt::lang_start_internal
             at /rustc/84c898d65adf2f39a5a98507f1fe0ce10a2b8dbc/library/std/src/rt.rs:148:20
  10: std::rt::lang_start
             at /rustc/84c898d65adf2f39a5a98507f1fe0ce10a2b8dbc/library/std/src/rt.rs:165:17
  11: main
  12: <unknown>
  13: __libc_start_main
  14: _start

Error: Process completed with exit code 1.
```

--- a/packages.txt
+++ b/packages.txt
@@ -1,0 +1,20 @@
+# List of packages to be published on `crates.io`, in this order.
+# - Specific arguments for cargo publish or cargo package may be added on the same line.
+# - See scripts/publish.sh and scripts/test_publish.sh for usage.
+linera-version --no-verify
+linera-witty-macros
+linera-witty
+linera-base
+linera-ethereum
+linera-views-derive
+linera-views
+linera-execution
+linera-chain
+linera-storage-service
+linera-storage
+linera-core
+linera-sdk-derive
+linera-sdk
+linera-rpc
+linera-service
+linera-service-graphql-client

--- a/packages.txt
+++ b/packages.txt
@@ -1,5 +1,5 @@
 # List of packages to be published on `crates.io`, in this order.
-# - Specific arguments for cargo publish or cargo package may be added on the same line.
+# - Specific arguments for cargo publish or cargo package may be added at the end, on the same line.
 # - See scripts/publish.sh and scripts/test_publish.sh for usage.
 linera-version --no-verify
 linera-witty-macros

--- a/scripts/publish.sh
+++ b/scripts/publish.sh
@@ -1,0 +1,10 @@
+#!/bin/bash
+
+set -x -e
+
+# Usage: grep -v '^#' packages.txt | scripts/publish.sh
+
+# Publish the given packages.
+while read LINE; do
+    cargo publish -p $LINE
+done

--- a/scripts/publish.sh
+++ b/scripts/publish.sh
@@ -1,8 +1,10 @@
 #!/bin/bash
 
-set -x -e
-
+# Publish crates from the current workspace.
+#
 # Usage: scripts/publish.sh packages.txt
+
+set -x -e
 
 # Publish the given packages.
 grep -v '^#' "$1" | while read LINE; do

--- a/scripts/publish.sh
+++ b/scripts/publish.sh
@@ -2,9 +2,9 @@
 
 set -x -e
 
-# Usage: grep -v '^#' packages.txt | scripts/publish.sh
+# Usage: scripts/publish.sh packages.txt
 
 # Publish the given packages.
-while read LINE; do
+grep -v '^#' "$1" | while read LINE; do
     cargo publish -p $LINE
 done

--- a/scripts/test_publish.sh
+++ b/scripts/test_publish.sh
@@ -1,13 +1,16 @@
 #!/bin/bash
 
-set -x -e
-
-# Requirements:
-#   cargo-local-registry
-#   cargo-index
-#   jq
+# Use a local registry to simulate publishing crates from the current workspace.
+#
+# Requirements (MacOS):
+#   cargo install cargo-local-registry
+#   cargo install cargo-index
+#   brew install jq
+#
 # Usage:
 #   scripts/test_publish.sh packages.txt REGISTRY
+
+set -x -e
 
 # Where to store the registry.
 mkdir -p "$2"
@@ -33,7 +36,7 @@ trap 'cd "$LINERA_DIR"; git checkout -f HEAD .cargo/config.toml' EXIT
 git init "$REGISTRY"/index || true
 (cd "$REGISTRY"/index; git add .; git commit -m 'update registry')
 
-# Build the packages in order.
+# Build the packages in order and add them to the local registry.
 grep -v '^#' "$1" | while read LINE; do
     ARGS=($LINE)
     CRATE="${ARGS[0]}"

--- a/scripts/test_publish.sh
+++ b/scripts/test_publish.sh
@@ -1,0 +1,40 @@
+#!/bin/bash
+
+set -x -e
+
+# Usage:
+#   cargo install cargo-local-registry
+#   cargo install cargo-index
+#   grep -v '^#' packages.txt | scripts/test_publish.sh REGISTRY
+
+# Where to store the registry.
+mkdir -p "$1"
+REGISTRY="$(cd "$1"; pwd)"
+shift
+
+# Make sure we're running from a clean repo.
+if [ ! -z "$(git status --porcelain)" ]; then
+    echo "The git repository has uncommitted changes"
+    exit 1
+fi
+
+# Synchronize the registry using `Cargo.lock`.
+(echo; echo '[source]') >> .cargo/config.toml
+cargo local-registry --git -s Cargo.lock "$REGISTRY" | tail -n +2 >> .cargo/config.toml
+
+echo "The following change was applied and should be reverted on exit:"
+git diff
+LINERA_DIR="$PWD"
+trap 'cd "$LINERA_DIR"; git checkout -f HEAD .cargo/config.toml' EXIT
+
+# Initialize the git repository for the index if needed. Ideally, we'd like to use `cargo
+# index init` first but the tool refuses to update an existing directory.
+git init "$REGISTRY"/index || true
+(cd "$REGISTRY"/index; git add .; git commit -m 'update registry')
+
+# Build the packages in order.
+while read LINE; do
+    ARGS=($LINE)
+    CRATE="${ARGS[0]}"
+    cargo index add --index "$REGISTRY"/index --upload "$REGISTRY" --index-url local --manifest-path "$CRATE"/Cargo.toml -- -p $LINE
+done

--- a/scripts/test_publish.sh
+++ b/scripts/test_publish.sh
@@ -2,9 +2,11 @@
 
 set -x -e
 
+# Requirements:
+#   cargo-local-registry
+#   cargo-index
+#   jq
 # Usage:
-#   cargo install cargo-local-registry
-#   cargo install cargo-index
 #   scripts/test_publish.sh packages.txt REGISTRY
 
 # Where to store the registry.
@@ -35,5 +37,6 @@ git init "$REGISTRY"/index || true
 grep -v '^#' "$1" | while read LINE; do
     ARGS=($LINE)
     CRATE="${ARGS[0]}"
-    cargo index add --index "$REGISTRY"/index --upload "$REGISTRY" --index-url local --manifest-path "$CRATE"/Cargo.toml -- -p $LINE
+    MANIFEST_PATH=$(cargo metadata --format-version 1 | jq -r '.packages[] | select(.name == "'$CRATE'") | .manifest_path')
+    cargo index add --index "$REGISTRY"/index --upload "$REGISTRY" --index-url local --manifest-path "$MANIFEST_PATH" -- -p $LINE
 done

--- a/scripts/test_publish.sh
+++ b/scripts/test_publish.sh
@@ -5,16 +5,15 @@ set -x -e
 # Usage:
 #   cargo install cargo-local-registry
 #   cargo install cargo-index
-#   grep -v '^#' packages.txt | scripts/test_publish.sh REGISTRY
+#   scripts/test_publish.sh packages.txt REGISTRY
 
 # Where to store the registry.
-mkdir -p "$1"
-REGISTRY="$(cd "$1"; pwd)"
-shift
+mkdir -p "$2"
+REGISTRY="$(cd "$2"; pwd)"
 
 # Make sure we're running from a clean repo.
-if [ ! -z "$(git status --porcelain)" ]; then
-    echo "The git repository has uncommitted changes"
+if [ ! -z "$(git status --porcelain .cargo/config.toml)" ]; then
+    echo "The file .cargo/config.toml has uncommitted changes"
     exit 1
 fi
 
@@ -22,8 +21,8 @@ fi
 (echo; echo '[source]') >> .cargo/config.toml
 cargo local-registry --git -s Cargo.lock "$REGISTRY" | tail -n +2 >> .cargo/config.toml
 
-echo "The following change was applied and should be reverted on exit:"
-git diff
+echo "The following change was applied to .cargo/config.toml and should be reverted on exit:"
+git diff | cat
 LINERA_DIR="$PWD"
 trap 'cd "$LINERA_DIR"; git checkout -f HEAD .cargo/config.toml' EXIT
 
@@ -33,7 +32,7 @@ git init "$REGISTRY"/index || true
 (cd "$REGISTRY"/index; git add .; git commit -m 'update registry')
 
 # Build the packages in order.
-while read LINE; do
+grep -v '^#' "$1" | while read LINE; do
     ARGS=($LINE)
     CRATE="${ARGS[0]}"
     cargo index add --index "$REGISTRY"/index --upload "$REGISTRY" --index-url local --manifest-path "$CRATE"/Cargo.toml -- -p $LINE

